README.md
# Laravel Prometheus Exporter

[![Latest Version on Packagist](https://img.shields.io/packagist/v/liveintent/laravel-prometheus-exporter.svg?style=flat-square)](https://packagist.org/packages/liveintent/laravel-prometheus-exporter)
[![GitHub Tests Action Status](https://img.shields.io/github/workflow/status/liveintent/laravel-prometheus-exporter/run-tests?label=tests)](https://github.com/liveintent/laravel-prometheus-exporter/actions?query=workflow%3ATests+branch%3Amaster)
[![GitHub Code Style Action Status](https://img.shields.io/github/workflow/status/liveintent/laravel-prometheus-exporter/Check%20&%20fix%20styling?label=code%20style)](https://github.com/liveintent/laravel-prometheus-exporter/actions?query=workflow%3A"Check+%26+fix+styling"+branch%3Amaster)
[![Total Downloads](https://img.shields.io/packagist/dt/liveintent/laravel-prometheus-exporter.svg?style=flat-square)](https://packagist.org/packages/liveintent/laravel-prometheus-exporter)


This is a Laravel package that automatically collects data and exposes a `/metrics` endpoint in your application which can then be scraped by [Prometheus](https://prometheus.io/).

For a full list of exported metrics, see [exporters](#exporters).

## Installation

You can install the package via composer:

```bash
composer require liveintent/laravel-prometheus-exporter
```

Once installed run the following command to generate the `metrics` config file. 

```bash
php artisan metrics:install
```

The package will auto-register itself.

## Usage

The `metrics` config file contains a list of enabled exporters. Each exporter will collect data and store it in your configured data store.

### Storage

The storage drivers currently supperted are redis, apc, and in-memory. You may adjust this value by setting the env `METRICS_STORAGE_DRIVER`.

The package uses the in-memory driver by default to help you get started, but you should change this as soon as you are ready as it's not useful for much beyond testing.

You will need the appropriate `pecl` extension installed ([apc](https://pecl.php.net/package/APCU) or [php-redis](https://pecl.php.net/package/redis)).

If you need to clear the storage, you may do so with:

```bash
php artisan metrics:clear
```

### Exporters

#### Request Duration Historam Exporter - `http_request_duration_seconds_bucket`

This will export histogram data for request duration. 

##### Example

```php
Exporters\RequestDurationHistogramExporter::class => [
    'enabled' => env('EXPORT_REQUEST_DURATION_HISTOGRAM', true),
    'options'
        'buckets' => [
            5, 10, 15, 20, 30, 40, 50, 60, 70, 80, 90, 100, 200
        ],
    ],
],
```

##### Labels

| name          | description            | example            |
|:--------------|:-----------------------|--------------------|
| service       | name of the service    | my-amazing-api     |
| environment   | the environment        | `qa`, `prod`, etc  |
| response_code | the http response code | `200`, `400`, etc  |
| method        | the http method        | `GET`, `POST`, etc |
| path          | the uri of the request | `/`, `/posts`, etc |

#### Request Memory Usage Historam Exporter - `http_request_memory_usage_bytes`

This will export histogram data for request memory usage. 

##### Example

```php
Exporters\RequestMemoryUsageHistogramExporter::class => [
    'enabled' => env('EXPORT_MEMORY_USAGE_HISTOGRAM', true),
    'options'
        'buckets' => [
            5, 10, 15, 20, 30, 40, 50, 60, 70, 80, 90, 100, 200
        ],
    ],
],
```

##### Labels

| name        | description            | example            |
|:------------|:-----------------------|--------------------|
| service     | name of the service    | my-amazing-api     |
| environment | the environment        | `qa`, `prod`, etc  |
| code        | the http response code | `200`, `400`, etc  |
| method      | the http method        | `GET`, `POST`, etc |
| path        | the uri of the request | `/`, `/posts`, etc |

#### Job Processing Time Historam Exporter - `job_process_time_seconds_bucket`

This will export histogram data for job execution duration. 

##### Example

```php
Exporters\JobProcessTimeHistogramExporter::class => [
    'enabled' => env('EXPORT_JOB_PROCESS_TIME_HISTOGRAM', true),
    'options'
        'buckets' => [
            0.1, .3, .5, .7, 1, 2, 3, 4, 5, 10, 30, 40, 60,
        ],
    ],
],
```

##### Labels

| name        | description                                | example                     |
|:------------|:-------------------------------------------|-----------------------------|
| service     | name of the service                        | my-amazing-api              |
| environment | the environment                            | `qa`, `prod`, etc           |
| name        | the classname of the job                   | `App\\Jobs\\ProcessPayment` |
| status      | the job status                             | `procesed`, `failed`        |

#### Job Wait Time Historam Exporter - `job_wait_time_seconds_bucket`

This will export histogram data for the time a job had to wait on the queue. 

**Note:** This exporter relies on [Laravel Horizon](https://laravel.com/docs/master/horizon).

##### Example

```php
Exporters\JobWaitTimeHistogramExporter::class => [
    'enabled' => env('EXPORT_JOB_WAIT_TIME_HISTOGRAM', true),
    'options'
        'buckets' => [
            0.1, .3, .5, .7, 1, 2, 3, 4, 5, 10, 30, 40, 60,
        ],
    ],
],
```

##### Labels

| name        | description                                | example                     |
|:------------|:-------------------------------------------|-----------------------------|
| service     | name of the service                        | my-amazing-api              |
| environment | the environment                            | `qa`, `prod`, etc           |
| name        | the classname of the job                   | `App\\Jobs\\ProcessPayment` |

#### Query Duration Historam Exporter - `db_query_time_seconds_bucket`

This will export histogram data for query execution times. 

##### Example

```php
Exporters\QueryDurationHistogramExporter::class => [
    'enabled' => env('EXPORT_QUERY_DURATION_HISTOGRAM', true),
    'options'
        'buckets' => [
            5, 10, 15, 20, 30, 40, 50, 60, 70, 80, 90, 100, 200
        ],
    ],
],
```

##### Labels

| name        | description         | example                    |
|:------------|:--------------------|----------------------------|
| service     | name of the service | my-amazing-api             |
| environment | the environment     | `qa`, `prod`, etc          |
| sql         | the sql query       | ``SELECT * FROM `users` `` |

### Writing New Exporters

You may also write your own exporter. You only need to implement two methods, `register` and `export`.

The following example increments a counter every time the `/` route is visited.

```php
<?php

namespace LiveIntent\LaravelPrometheusExporter\Exporters;

use Illuminate\Foundation\Http\Events\RequestHandled;

class HomePageVisitsCountExporter extends Exporter
{
    /**
     * Register the watcher.
     *
     * @return void
     */
    public function register()
    {
        $this->app['events']->listen(RequestHandled::class, [$this, 'export']);
    }

    /**
     * Export metrics.
     *
     * @param  \Illuminate\Foundation\Http\Events\RequestHandled  $event
     * @return void
     */
    public function export($event)
    {
        $uri =  str_replace($event->request->root(), '', $event->request->fullUrl()) ?: '/';
        
        if ($uri === '/') {
            $counter = $this->registry->getOrRegisterCounter(
                '', 'home_page_visits_counter', 'it is a silly example'
            );

            $counter->inc();
        }
    }
}
```

Register your Exporter class by placing it in the list of exporters found in the `metrics` config file.


CHANGELOG.md
# Changelog

All notable changes to `laravel-prometheus-exporter` will be documented in this file.

Everything in v1 is subject to breaking changes while we work out the kinks.

## 1.4.0 - 2021-04-22

- go back to using horizon
- account for batched jobs in wait time

## 1.3.0 - 2021-04-21

- use our own pushedAt timestamp via Cache so as not to rely on horizon
- normalize route parameters so each new value is not considered a separate route

## 1.2.0 - 2021-04-16

- added job wait time exporter
- renamed job duration exporter class to job process time exporter

## 1.1.0 - 2021-04-14

- renamed metrics to comply with prometheus best practices
- added ability to opt out of tracking certain queries and paths

## 1.0.1 - 2021-04-02

- actual initial release ;)

## 1.0.0 - 2021-04-02

- initial release


config/metrics.php
<?php

use LiveIntent\LaravelPrometheusExporter\Exporters;

return [

    /*
    |--------------------------------------------------------------------------
    | Metrics Domain
    |--------------------------------------------------------------------------
    |
    | This is the subdomain where the metrics will be accessible from. If the
    | setting is null, the metrics will reside under the same domain as the
    | application. Otherwise, this value would be used for the subdomain.
    |
    */

    'domain' => env('METRICS_DOMAIN', null),

    /*
    |--------------------------------------------------------------------------
    | Metrics Path
    |--------------------------------------------------------------------------
    |
    | This is the URI path where the metrics will be accessible from. You may
    | change this path to anything you like, but by default Prometheus will
    | be looking for '/metrics', so you probably don't want to change it.
    |
    */

    'path' => env('METRICS_PATH', 'metrics'),

    /*
    |--------------------------------------------------------------------------
    | Metrics Storage Driver
    |--------------------------------------------------------------------------
    |
    | This configuration option determines the storage driver that'll be used
    | to store metric data. In most cases, you will want something that can
    | persist state across multiple Laravel instances and is fast enough.
    |
    | Supported drivers: "apc", "redis", "memory"
    */

    'driver' => env('METRICS_STORAGE_DRIVER', 'memory'),

    'storage' => [
        'redis' => [
            'host' => env('REDIS_HOST', '127.0.0.1'),
            'port' => env('REDIS_PORT', '6379'),
            'password' => env('REDIS_PASSWORD', null),
            'timeout' => 0.1,
            'read_timeout' => '10',
            'persistent_connections' => null,
            'prefix' => env('REDIS_METRICS_PREFIX', null),
        ],
    ],

    /*
    |--------------------------------------------------------------------------
    | Metrics Exporter Primary Switch
    |--------------------------------------------------------------------------
    |
    | This primary switch may be used as an override to turn off every metric
    | exporter regardless of their individual configuration, which is meant
    | to be a convenient way to enable or disable metric data collection.
    |
    */

    'enabled' => env('METRICS_EXPORTER_ENABLED', true),

    /*
    |--------------------------------------------------------------------------
    | Metric Exporters
    |--------------------------------------------------------------------------
    |
    | The following array lists the "exporters" that will be registered. Each
    | exporter takes the data captured by Laravel and makes it available to
    | the metrics endpoint to expose. Prometheus can then store the data.
    |
    | You may also define additional custom exporters and add them below.
    |
    */

    'exporters' => [

        /*
        |--------------------------------------------------------------------------
        | HTTP Request Exporters
        |--------------------------------------------------------------------------
        */

        Exporters\RequestDurationHistogramExporter::class => [
            'enabled' => env('EXPORT_REQUEST_DURATION_HISTOGRAM', true),
            'options' => [
                'buckets' => [
                    0.01, 0.02, 0.03, 0.04, 0.05, .06, 1, 1.2, 1.5, 1.8, 2, 3, 4, 5, 10
                ],
                'ignore_path_regex' => [
                    '/telescope/'
                ]
            ],
        ],

        Exporters\RequestMemoryUsageHistogramExporter::class => [
            'enabled' => env('EXPORT_REQUEST_MEMORY_USAGE_HISTOGRAM', true),
            'options' => [
                'buckets' => [
                    500000, 1000000, 2000000, 3000000, 4000000, 5000000, 6000000, 7000000, 8000000, 9000000, 10000000, 30000000
                ],
                'ignore_path_regex' => [
                    '/telescope/'
                ]
            ],
        ],

        /*
        |--------------------------------------------------------------------------
        | Job Exporters
        |--------------------------------------------------------------------------
        */

        Exporters\JobProcessTimeHistogramExporter::class => [
            'enabled' => env('EXPORT_JOB_PROCESS_TIME_HISTOGRAM', true),
            'options' => [
                'buckets' => [
                    0.01, 0.02, 0.03, 0.04, 0.05, .06, 1, 1.2, 1.5, 1.8, 2, 3, 4, 5, 10
                ],
            ],
        ],

        Exporters\JobWaitTimeHistogramExporter::class => [
            'enabled' => env('EXPORT_JOB_WAIT_TIME_HISTOGRAM', true),
            'options' => [
                'buckets' => [
                    0.1, .3, .5, .7, 1, 2, 3, 4, 5, 10, 30, 40, 60,
                ],
                'cache' => [
                    'store' => 'redis',
                    'expiration' => now()->addDay()
                ]
            ],
        ],

        /*
        |--------------------------------------------------------------------------
        | Query Exporters
        |--------------------------------------------------------------------------
        */

        Exporters\QueryDurationHistogramExporter::class => [
            'enabled' => env('EXPORT_QUERY_DURATION_HISTOGRAM', true),
            'options' => [
                'buckets' => [
                    .001, .005, 0.01, 0.02, 0.03, 0.04, 0.05, .06, 1, 1.2, 1.5, 1.8, 2, 3, 4, 5, 10
                ],
                'ignore_query_regex' => [
                    '/telescope_entries/',
                    '/create\s+table/',
                    '/alter\s+table/',
                    '/drop\s+table/',
                ]
            ],
        ],
    ],
];
